watertap/tools/oli_api/flash.py

```diff
@@ -0,0 +1,494 @@
+
+        :param use_scaling_rigorous: boolean indicating desired state of 'rigorous' and 'estimated' optional properties
+        """
+
+        if bool(use_scaling_rigorous) == bool(
+            self.optional_properties["prescalingTendenciesRigorous"]
+        ):
+            return
+        new_values = {
+            k: not v for k, v in self.optional_properties.items() if "prescaling" in k
+        }
+        self.optional_properties.update(new_values)
+
+    def build_flash_calculation_input(
+        self,
+        flash_method,
+        state_vars,
+        water_analysis_output=None,
+        inflows_phase="total",
+        use_scaling_rigorous=True,
+    ):
+        """
+        Builds a base dictionary required for OLI flash analysis.
+
+        :param flash_method: string name of OLI flash flash_method to use
+        :param state_vars: dictionary containing solutes, temperatures, pressure, and units
+        :param water_analysis_output: dictionary to extract inflows from (required if not wateranalysis flash)
+        :param inflows_phase: string indicating desired phase of inflows
+        :param use_scaling_rigorous: boolean switch to use estimated or rigorous solving for prescaling metrics
+
+        :return inputs: dictionary containing inputs for specified OLI flash analysis
+        """
+
+        inputs = {"params": {}}
+
+        if flash_method == "wateranalysis":
+            inputs["params"] = {
+                "waterAnalysisInputs": self._build_input_list(state_vars)
+            }
+        else:
+            if not water_analysis_output:
+                raise IOError(
+                    "Run wateranalysis flash to generate water_analysis_output data."
+                )
+
+            inputs["params"] = {
+                "temperature": {
+                    "unit": str(state_vars["units"]["temperature"]),
+                    "value": float(state_vars["temperature"]),
+                },
+                "pressure": {
+                    "unit": str(state_vars["units"]["pressure"]),
+                    "value": float(state_vars["pressure"]),
+                },
+                "inflows": self._extract_inflows(water_analysis_output, inflows_phase),
+            }
+
+        self._set_prescaling_calculation_mode(use_scaling_rigorous)
+        inputs["params"].update({"optionalProperties": dict(self.optional_properties)})
+        inputs["params"].update({"unitSetInfo": dict(self.output_unit_set)})
+        return inputs
+
+    def _extract_inflows(self, water_analysis_output, inflows_phase):
+        """
+        Extract molecular concentrations from OLI flash output.
+
+        :param water_analysis_output: stream output from OLI flash calculation
+        :param inflows_phase: string indicating desired phase of inflows
+
+        :return inflows: dictionary containing molecular concentrations
+        """
+
+        if inflows_phase in ["liquid1", "vapor"]:
```

Comment:
Another thing to contemplate for the future---I think liquid2 is another possible phase in OLI. I can see this _**potentially**_ being useful for our upcoming work on solvent extraction

Reply:
(which would ideally make use of two liquids in a more detailed representation of the required property model)

Reply:
That's true, probably along with unlocking the other flash calculations